feat: Bootstrap Makefile

charlielye opened the pull request:
Get some extra build performance with a better defined dependency tree specified with a Makefile.

This first pass goes more granular with:
* avm transpiler
* bb targets
* l1 contracts

We now start the "test engine" before the build. As the build graph progresses and tests become available to run, we write them to `/tmp/test_cmds` which the test engine is reading from. This means we get a head-start on running tests where we can. We can also do the "yarn-project building bench" phase while tests are running, rather than after.

This solidifies conceptually, the modes in which CI runs:
* Local bootstrap CI_FULL=0 - `./bootstrap.sh`
* Regular prs are x86 CI_FULL=0 + tests - `./bootstrap.sh ci-fast`
* x86 merge queue is CI_FULL=1 + tests + benches - `./bootstrap.sh ci-full`
* arm merge queue is CI_FULL=0 + tests - `./bootstrap ci-fast`
* releases are CI_FULL=0 + semver check passes (no tests). `./bootstrap.sh release`

Within bootstrap.sh scripts you should really only ever need conditionals on CI_FULL.
We should avoid architecture dependent conditionals, semver checks, etc.

Introduces a `ci-release-pr` label. Put it on your PR and it'll go through a full release with tag e.g. `v0.0.1-commit-<hash>`

Still to todo:
* A major bottleneck right now is compiling the protocol circuits which prevents all downstream stages while waiting on a couple of checkpoint circuits. Consider better handling bb vs bb-avm. Noir projects does not need to wait on bb-avm (edit: apparently one circuit does - still, consider granularity, and can we just "link in" the result later in the pipeline).
* Go more granular with the yarn-project build, which should allow later stages to start building sooner (e.g. while type-checking is being performed).
* AND/OR just switch tsgo.

**Time New:**
`CPUS=128 NO_CACHE=1 NO_SPOT=1 ci full = 44m`
`CPUS=192 NO_CACHE=1 NO_SPOT=1 ci full = 39m`

charlielye commented:
<!-- CI3_FLAKE_DETECTION_COMMENT -->
## Flakey Tests

🤖 says: This CI run detected **1** tests that failed, but were tolerated due to a .test_patterns.yml entry.

```
\033[38;2;188;109;208mFLAKED\033[0m (\033[38;2;250;217;121m8;;http://ci.aztec-labs.com/4b3afbd0c0e3128e4b3afbd0c0e3128e8;;\033[0m): BOX=vanilla BROWSER=* run_compose_test vanilla-all-browsers box boxes (367s) (code: 1) (\033[38;2;188;109;208mAztec Bot\033[0m: feat: Bootstrap Makefile (#18135))
```

ludamad commented on Makefile:
```diff
@@ -0,0 +1,318 @@
+# project-path: Path to the project directory
+# function-name: Optional bootstrap.sh command (defaults to $(BUILD_MODE))
+define build
+	$(call run_command,$(1),$(ROOT)/$(2),$(ROOT)/ci3/denoise './bootstrap.sh $(if $(3),$(3),$(BUILD_MODE))')
```
my only concern is that the makefile syntax is harder to read than BASH :)

charlielye replied:
but it isn't!
It's beautifully declarative and does away of crazy logic everywhere in favour of 3 or 4 clean targets.

ludamad replied:
```
define build
	$(call run_command,$(1),$(ROOT)/$(2),$(ROOT)/ci3/denoise './bootstrap.sh $(if $(3),$(3),$(BUILD_MODE))')
```
I was commenting on this part. The dependency part of Makefile's is perfectly clean

ludamad replied:
e.g. `$(if $(3),$(3),$(BUILD_MODE))` is a mystery to me at a glance

charlielye replied:
I've just pushed a commit to remove this anyway, as it wasn't needed.
I've removed the fast/full tasks from all bootstrap scripts except root.
Only a couple had different paths, which can be handled by checking CI_FULL var.

ludamad reviewed (approved):
Overall very clean other than Makefile syntax itself being a bit rough. LGTM!

charlielye committed:
feat: Bootstrap Makefile

Get some extra build performance with a better defined dependency tree specified with a Makefile.

This first pass goes more granular with:
* avm transpiler
* bb targets
* l1 contracts

We now start the "test engine" before the build. As the build graph progresses and tests become available to run, we write them to `/tmp/test_cmds` which the test engine is reading from. This means we get a head-start on running tests where we can. We can also do the "yarn-project building bench" phase while tests are running, rather than after.

This solidifies conceptually, the modes in which CI runs:
* Local bootstrap CI_FULL=0 - `./bootstrap.sh`
* Regular prs are x86 CI_FULL=0 + tests - `./bootstrap.sh ci-fast`
* x86 merge queue is CI_FULL=1 + tests + benches - `./bootstrap.sh ci-full`
* arm merge queue is CI_FULL=0 + tests - `./bootstrap ci-fast`
* releases are CI_FULL=0 + semver check passes (no tests). `./bootstrap.sh release`

Within bootstrap.sh scripts you should really only ever need conditionals on CI_FULL.
We should avoid architecture dependent conditionals, semver checks, etc.

Introduces a `ci-release-pr` label. Put it on your PR and it'll go through a full release with tag e.g. `v0.0.1-commit-<hash>`

Still to todo:
* A major bottleneck right now is compiling the protocol circuits which prevents all downstream stages while waiting on a couple of checkpoint circuits. Consider better handling bb vs bb-avm. Noir projects does not need to wait on bb-avm (edit: apparently one circuit does - still, consider granularity, and can we just "link in" the result later in the pipeline).
* Go more granular with the yarn-project build, which should allow later stages to start building sooner (e.g. while type-checking is being performed).
* AND/OR just switch tsgo.

**Time New:**
`CPUS=128 NO_CACHE=1 NO_SPOT=1 ci full = 44m`
`CPUS=192 NO_CACHE=1 NO_SPOT=1 ci full = 39m`
